rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,199 @@
+            num_partitions,
+            pq_width,
+            output_dir,
+            schema,
+        })
+    }
+
+    pub async fn dump_unsorted_stream(
+        &self,
+        data: impl RecordBatchStream + Unpin + 'static,
+    ) -> Result<()> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let writer = object_store.create(&path).await?;
+
+        let mut file_writer =
+            FileWriter::with_object_writer(writer, self.schema.clone(), &Default::default())?;
+
+        let mut data = Box::pin(data);
+
+        while let Some(batch) = data.next().await {
+            file_writer.write(&[batch?]).await?;
+        }
+
+        file_writer.finish().await?;
+
+        Ok(())
+    }
+
+    pub async fn count_partition_size(&self) -> Result<Vec<u64>> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        let mut partition_sizes = vec![0; self.num_partitions as usize];
+
+        let lance_schema = reader
+            .schema()
+            .project(&[PART_ID_COLUMN])
+            .expect("part id should exist");
+
+        let mut stream = stream::iter(0..reader.num_batches())
+            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, &lance_schema))
+            .buffered(64);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+            let part_ids: &UInt32Array = as_primitive_array(batch.column(0));
+            for part_id in part_ids.values() {
+                partition_sizes[*part_id as usize] += 1;
+            }
+        }
+
+        Ok(partition_sizes)
+    }
+
+    pub async fn shuffle_to_partitions(
+        &self,
+        partition_size: Vec<u64>,
+    ) -> Result<(Vec<Vec<u64>>, Vec<Vec<u8>>)> {
+        let mut row_id_buffers = partition_size
+            .iter()
+            .map(|s| Vec::with_capacity(*s as usize))
+            .collect::<Vec<_>>();
+        let mut pq_code_buffers = partition_size
+            .iter()
+            .map(|s| Vec::with_capacity((*s as usize) * self.pq_width))
+            .collect::<Vec<_>>();
+
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        info!("Shuffling into memory");
+        let mut bar = kdam::BarBuilder::default()
+            .total(reader.num_batches())
+            .build()
+            .unwrap();
+
+        let mut stream = stream::iter(0..reader.num_batches())
+            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, reader.schema()))
+            .buffered(16);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+
+            let row_ids = as_primitive_array::<UInt64Type>(
+                batch
```

Comment:
`batch.column_by_name().expect().as_primitive_array()` seems to be the endorsed approach in arrow-rs

Reply:
I don't think there is an extension for this.